mycore-base/src/main/java/org/mycore/datamodel/common/MCRFileBaseCacheObjectIDGenerator.java

```diff
@@ -0,0 +1,173 @@
+package org.mycore.datamodel.common;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.mycore.common.MCRException;
+import org.mycore.common.MCRUtils;
+import org.mycore.common.config.MCRConfiguration2;
+import org.mycore.datamodel.metadata.MCRObjectID;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class MCRFileBaseCacheObjectIDGenerator implements MCRObjectIDGenerator {
+
+    private static final Logger LOGGER = LogManager.getLogger();
+
+    static ConcurrentHashMap<String, ReentrantReadWriteLock> locks = new ConcurrentHashMap<>();
+
+    private static Path getCacheFilePath(String baseId) {
+
+        Path dataDir = getDataDirPath();
+
+        Path idCachePath = dataDir.resolve("id_cache");
+        if (!Files.exists(idCachePath)) {
+            synchronized (MCRFileBaseCacheObjectIDGenerator.class) {
+                if (!Files.exists(idCachePath)) {
+                    try {
+                        Files.createDirectory(idCachePath);
+                    } catch (IOException e) {
+                        throw new MCRException(
+                            "Could not create " + idCachePath.toAbsolutePath() + " directory", e);
+                    }
+                }
+            }
+        }
+
+        Path cacheFile = MCRUtils.safeResolve(idCachePath, baseId);
+        if (!Files.exists(cacheFile)) {
+            synchronized (MCRFileBaseCacheObjectIDGenerator.class) {
+                if (!Files.exists(cacheFile)) {
+                    try {
+                        Files.createFile(cacheFile);
+                    } catch (IOException e) {
+                        throw new MCRException("Could not create " + cacheFile.toAbsolutePath(), e);
+                    }
+                }
+            }
+        }
+        return cacheFile;
+    }
+
+    static Path getDataDirPath() {
+        Path path = Paths.get(MCRConfiguration2.getStringOrThrow("MCR.datadir"));
+        if (Files.exists(path) && !Files.isDirectory(path)) {
+            throw new MCRException("Data directory does not exist or is not a directory: " + path);
+        }
+        return path;
+    }
+
+    private static void writeNewID(MCRObjectID nextID, ByteBuffer buffer, FileChannel channel, Path cacheFile)
+        throws IOException {
+        buffer.clear();
+        channel.position(0);
+        buffer.putInt(nextID.getNumberAsInteger());
```

Comment:
For me if would be nicer to see the Integer not as byte, but readable string in the file - or maybe store the whole MCRObjectID-String?
So it would be easier to Debug the MyCoRe-System.

Reply:
i store the entire id as string now